text/0171-cloudfront-redesign.md

```diff
@@ -0,0 +1,477 @@
+
+# Detailed Design
+
+The design creates one new resource (`Distribution`) and two new classes (`Origin` and `Behavior`) to replace the current
+`CloudFrontWebDistribution` construct. Each of these new classes comes with helper methods (e.g., `addBehavior`) to make assembling more complex
+distributions easier, as well as factory constructors to make it easier to build common Origin and Behavior patterns (e.g., `Origin.fromBucket`).
+
+The following is an incomplete, but representative, listing of the API:
+
+```ts
+class Distribution extends BaseDistribution {
+  static fromArn(scope: Construct, id: string, distributionArn: string): IDistribution;
+  static fromAttributes(scope: Construct, id: string, distributionArn: string): IDistribution;
+
+  constructor(scope: Construct, id: string, props: DistributionProps) {}
+
+  addOrigin(options: OriginOptions): Origin
+  addBehavior(pathPattern: string, options: BehaviorOptions): Behavior
+}
+
+class Origin {
+  static fromBucket(bucket: s3.IBucket, behaviorOptions?: BehaviorProps): Origin
+  static fromWebsiteBucket(bucket: s3.IBucket, behaviorOptions?: BehaviorProps): Origin
+  static fromLoadBalancerV2(loadBalancer: elbv2.ApplicationLoadBalancer, behaviorOptions?: BehaviorProps): Origin
+  static fromHttpServer(options: ServerOriginOptions, behaviorOptions?: BehaviorProps): Origin
+
+  constructor(props: OriginProps) {}
+
+  addBehavior(pathPattern: string, options: BehaviorOptions): Behavior
+}
+
+class Behavior {
+  constructor(props: BehaviorProps) {}
+
+  addFunctionAssociation(options: FunctionAssociationOptions): FunctionAssociation
+}
+```
+
+## Nested structure and relationships
+
+The `Distribution` has one top-level origin and behavior, which aligns to how the vast majority of customers use CloudFront today (based on public
+CDK examples). Customers can add additional origins (and origin groups) and behaviors, which are modeled as `additionalOrigins` and
+`additionalBehaviors`, respectively. Each origin may have multiple behaviors associated with it, so the relationship is modeled such that behaviors
+are added to origins. However, an authoritative list of behaviors is kept on the distribution to preserve ordering. This is done similarly to how
+ECS clusters keep track of Fargate profiles as they are created and associated with the cluster. In the below example, the '/api/\*' behavior for
+the load balancer origin will be ordered first, then the '/api/errors/\*' behavior on the bucket.
+
+```ts
+const myWebsiteBucket = new s3.Bucket(...);
+const myMultiOriginDistribution = new Distribution(this, 'myDist', {
+  origin: Origin.fromWebsiteBucket(this, 'myOrigin', myBucket),
+  additionalOrigins: [Origin.fromLoadBalancerV2(this, 'myOrigin', myLoadBalancer, {
+    pathPattern: '/api/*',
+    allowedMethods: AllowedMethods.ALL,
+    forwardQueryString: true,
+  })];
+});
+myMultiOriginDistribution.origin.addBehavior('/api/errors/*', ...);
```

Comment:
This (the fact that you have behaviors both on Distribution and Origin) wrecks my poor little brain 😕. Any chance we can simplify it, so that you can only add Behavior to Origin?

Reply:
So this is … tricky. To create a distribution, there must be one origin and a default behavior; this default behavior must have a pathPattern of “*”. In order for any additional origins to receive traffic, there must be a new behavior with a pathPattern defined; so while not strictly required, it makes sense generally for each origin to have a behavior associated with it at creation. There are exceptions to this which I’ll get back to.

So the obvious thing would be to always have the `Origin.from*` methods create a behavior to associate with the origin, which would require a pathPattern. But for the default origin/behavior, those should be optional. So one option would be to alter the factory methods to look like:

```ts
static fromBucket(bucket: s3.IBucket, pathPattern?: string, cacheBehaviorProps?: BehaviorProps): Origin
```

For the “default” origin, the pathPattern and cacheBehaviorProps can be omitted, but for all others they would be required. This would have to be enforced by the Distribution when the origin is added (if an “additionalOrigin” with a “*” behavior is added, throw an error).

There are two caveats with this approach: (1) Behavior creation order can influence the origin routing (see detailed design and rationale for more details). Implicitly creating behaviors may make this approach more confusing. (2) Origin groups — with origin groups, only one behavior is needed which points to the primary origin, and the fallback origin is routed to only in case of failure of the primary. In that case, the auto-created behavior of the fallback origin would be ignored.

I’m going to call in @iliapolo for his feedback on this one as well.

Reply:
Before I chip in, let me ask 2 clarifying question.

I understand the default origin always has to have the `pathPattern` equal to `"*"` (so ideally, we don't want to burden the customer with providing it). But:

1. Can the default behavior have a custom cache behavior?
2. Can the non-default behaviors skip providing the cache behavior part?

Reply:
> 1. Can the default behavior have a custom cache behavior?

Yes, you can customize the default cache behavior the exact same as any other behavior.

> 2. Can the non-default origins skip providing the cache behavior part?

Yes, except for the 'pathPattern' part. So you can define a second origin and accept all of the default behaviors, and only specify what path redirects to the second origin. So this would be valid:

```ts
mySecondOrigin.addBehavior('/images/*.jpg');
```

This makes sense for the first behavior on any additional origin; however, adding a second behavior on an origin _without_ customization wouldn't make sense.

Reply:
Yeah, you discussion aligns with some of the comments I made along the way...

Its a thinker. 

I definitely think its worth coming up with a way to move behavior declaration to the origin level only. As far as I can tell, the problem is the `pathPattern`, which is fixed for the default origin, but configurable for the additional origins. Intuitively, the API I would expect is:

```typescript
const myWebDistribution = new cloudfront.Distribution(this, 'myDist', {
  origin: cloudfront.Origin.fromHTTPServer({
    domainName: 'www.example.com',
    originProtocolPolicy: OriginProtocolPolicy.HTTPS_ONLY,
    behavior: {
      allowedMethods: AllowedMethods.ALL,
      viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
    }
  }),
});
```

Which will use `*` as the `pathPattern`. Then, configure additional origins using:

```typescript
myWebDistribution.addOrigin('/api/*', Origin.fromBucket(myBucket, {
  allowedMethods: AllowedMethods.ALL,
  forwardQueryString: true,
}));
```

Basically, removing `pathPattern` from `BehaviorProps`.  Can this work? 

For the constructor, we can have:

```typescript
const myWebDistribution = new cloudfront.Distribution(this, 'myDist', {
  origin: cloudfront.Origin.fromHTTPServer({
    domainName: 'www.example.com',
    originProtocolPolicy: OriginProtocolPolicy.HTTPS_ONLY,
    behavior: {
      allowedMethods: AllowedMethods.ALL,
      viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
    }
  }),
  additionalOrigins: {
    '/api/*': Origin.fromBucket(myBucket, {allowedMethods: AllowedMethods.ALL, forwardQueryString: true});
  }
});
```

One thing I don't understand is, does it make sense to have multiple behavior on the same origin? As far as I can tell, a behavior is configured to a path, which cloudfront routes to some origin. Right?

Reply:
> One thing I don't understand is, does it make sense to have multiple behavior on the same origin? As far as I can tell, a behavior is configured to a path, which cloudfront routes to some origin. Right?

Yes, having multiple behaviors on the same origin does make sense. The behavior does route to an origin, but also has all of the controls around caching, headers, forwarded values, etc. You might have one behavior on an origin for longer-cached, mostly static content and another for more dynamic content that needs query parameters and cookies.

Reply:
Can you post an elaborate example of the current API with multiple origins and multiple behaviors per origin?

Reply:
I'd also like to see a complex example.

To reiterate my previous comment, I think I'd like to see the distinction between primary and additional origins disappear.

Reply:
A complex example already exists in the "Comparison" section (Use case 3). 

After thinking about this a bit, I think the simplest way to resolve this is to remove the `behavior` property from the Distribution to model the default behavior. This aligns with @skinny85 's original comment of making it so behaviors can only be attached to Origins. I've updated the RFC and examples to match this.

*Implementation notes:* What this means is that all of the `Origin.from*` methods accept the BehaviorProps (as they already did) as an optional argument. If no BehaviorProps are provided, then the Origin will originally not be assigned a Behavior. At Distribution creation, a default behavior will be created for the default origin, and an error will be thrown for any other origin without a behavior (as no sane default exists). This is a reasonably minor change to the design -- as evidenced by how easy it was to make the change to the RFC -- but I think resolves the open issues here. Lazy-creating/binding the behaviors solves my concern about origin groups having redundant behaviors, and ordering isn't a strict concern because the default behavior is always last anyway.

Reply:
> I'd also like to see a complex example.
> 
> To reiterate my previous comment, I think I'd like to see the distinction between primary and additional origins disappear.

But there is a very important distinction between primary and secondary origins, right?

A secondary origin must have a behavior associated with it with some `pathPattern`. While a primary origin has a default behavior (= with `pathPatter` = `*`) associated with it automatically.